x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamInput.java

```diff
@@ -60,6 +71,8 @@ public NameId apply(long streamNameId) {
 
     private static final Supplier<LongFunction<NameId>> DEFAULT_NAME_ID_FUNC = NameIdMapper::new;
 
+    private final Map<Integer, Block> cachedBlocks = new HashMap<>();
```

Comment:
Let's decouple this from the stream, similar to `NameIdMapper` - this way we could support both streams that don't do any caching and followed by those who do.
How common is it that we'll have duplicate blocks? Comparing their content will take CPU and I wonder if being opportunistic about reusing them (when we know it's the same data) isn't an alternative for at least some cases.

Reply:
In the plan I think we're fairly likely to have duplicates - at least, that's how I want to build the serialization. Let the streaming deal with the duplicates.